utils/filesystem/filepath.go

```diff
@@ -73,15 +73,145 @@ func FilePathBase(fs FS, fp string) string {
+	return FilePathFromPlatformPathSeparator(fs, filepath.Base(FilePathToPlatformPathSeparator(fs, fp)))
+}
+
+// FilePathDir has the same behaviour as filepath.Dir, but can handle different filesystems.
+func FilePathDir(fs FS, fp string) string {
+	if fs == nil {
+		return ""
+	}
+
+	if isSlashSeparator(fs) {
+		return path.Dir(fp)
+	}
+	return FilePathFromPlatformPathSeparator(fs, filepath.Dir(FilePathToPlatformPathSeparator(fs, fp)))
+}
+
+// FilePathIsAbs has the same behaviour as filepath.IsAbs, but can handle different filesystems.
+func FilePathIsAbs(fs FS, fp string) bool {
+	if fs == nil {
+		return false
+	}
+	if isSlashSeparator(fs) {
+		return path.IsAbs(fp)
+	}
+	return filepath.IsAbs(FilePathToPlatformPathSeparator(fs, fp))
+}
+
+// FilePathRel has the same behaviour as filepath.FilePathRel, but can handle different filesystems.
+func FilePathRel(fs FS, basepath, targpath string) (rel string, err error) {
+	if fs == nil {
+		err = commonerrors.UndefinedVariable("filesystem specification")
+		return
+	}
+	rel, err = filepath.Rel(FilePathToPlatformPathSeparator(fs, basepath), FilePathToPlatformPathSeparator(fs, targpath))
+	if err != nil {
+		return
+	}
+	rel = FilePathFromPlatformPathSeparator(fs, rel)
+	return
+}
+
+// FilePathSplit has the same behaviour as filepath.Split, but can handle different filesystems
+func FilePathSplit(fs FS, fp string) (dir, file string) {
+	if fs == nil {
+		return
+	}
+
+	if isSlashSeparator(fs) {
+		return path.Split(fp)
+	}
+	dir, file = filepath.Split(FilePathToPlatformPathSeparator(fs, fp))
+	dir = FilePathFromPlatformPathSeparator(fs, dir)
+	return
+}
+
+// FilePathAbs tries to be similar to filepath.Abs behaviour but without using platform information or location.
+func FilePathAbs(fs FS, fp, currentDirectory string) string {
+	if FilePathIsAbs(fs, fp) {
+		return FilePathClean(fs, fp)
+	}
+	return FilePathJoin(fs, currentDirectory, fp)
+}
+
+// FilePathToPlatformPathSeparator returns the result of replacing each separator character in path with a platform path separator character
+func FilePathToPlatformPathSeparator(fs FS, path string) string {
+	if fs.PathSeparator() == platform.PathSeparator {
+		return path
+	}
+	return strings.ReplaceAll(path, string(fs.PathSeparator()), string(platform.PathSeparator))
+}
+
+// FilePathFromPlatformPathSeparator returns the result of replacing each platform path separator character in path with filesystem's path separator character
+func FilePathFromPlatformPathSeparator(fs FS, path string) string {
+	if fs.PathSeparator() == platform.PathSeparator {
+		return path
 	}
+	return strings.ReplaceAll(path, string(platform.PathSeparator), string(fs.PathSeparator()))
+}
 
-	return filepath.Base(fp)
+// FilePathToSlash is filepath.ToSlash  but using filesystem path separator rather than platform's
+func FilePathToSlash(fs FS, path string) string {
+	if isSlashSeparator(fs) {
+		return path
+	}
+	return strings.ReplaceAll(path, string(fs.PathSeparator()), "/")
```

Comment:
What if it is an absolute path? Sometimes we do filepath.Join("/", "a", "b", "c") etc. Which would not work on windows, or we might pass an absolute path.

I suppose that if you're working with absolute paths then the code is already platform specific anyway so maybe it is unneeded. In that case I think adding that information to the docstring would be good to make it clear you can't supply absolute paths.

There are utilities in here for working with absolute paths though that I saw whilst reviewing. If you do add checks for absolute paths then can you add a test in `TestFilePathFromTo`

Reply:
create a story for this